### PR TITLE
Improve Lsra stats

### PR DESCRIPTION
**Pull request by kunalspathak**
Improve Lsra stats to include register selection heuristics information. The sample output looks like this:

```cmd
----------
LSRA Stats : System.Runtime.CompilerServices.CastHelpers:LdelemaRef(System.Array,int,long):byref
----------
Total Tracked Vars:  5
Total Reg Cand Vars: 5
Total number of Intervals: 8
Total number of RefPositions: 39
Total Number of spill temps created: 0
..........
BB01 [  100.00]: COVERS = 2, RELATED_PREFERENCE = 1, BEST_FIT = 1, REG_ORDER = 2
..........
Total SpillCount : 0   Weighted: 0.000000
Total CopyReg : 0   Weighted: 0.000000
Total ResolutionMovs : 0   Weighted: 0.000000
Total SplitEdges : 0   Weighted: 0.000000
..........
Total COVERS [# 4] : 2   Weighted: 200.000000
Total RELATED_PREFERENCE [# 7] : 1   Weighted: 100.000000
Total BEST_FIT [#11] : 1   Weighted: 100.000000
Total REG_ORDER [#13] : 2   Weighted: 200.000000
```

I have also added a way to store the stats in csv form if we set `COMPlus_JitLsraStats=2`. With that, I can easily summarize graphically which heuristics is popular. Below is the data I gathered by running `MulMatrix` microbenchmark and all the register selection heuristics during that:

image

### CODE REVIEW

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib

File: src/coreclr/jit/lsra.cpp [hunk 1]
```diff
@@ -3075,6 +3086,10 @@ regNumber LinearScan::allocateReg(Interval* currentInterval, RefPosition* refPos
     else if (!found)
     {
         found = selector.applySelection(FREE, freeCandidates);
+        if (found)
+        {
+            INTRACK_STATS(updateLsraStat(LsraStat::REGSEL_FREE, refPosition->bbNum));
```
Comment (BruceForstall) [inline src/coreclr/jit/lsra.cpp:3091]: This is such a common pattern, would it be useful to have a macro
```
#define INTRACK_STATS_IF(condition,work) if (condition) work;
```
?

and:
```
INTRACK_STATS_IF(found, updateLsraStat(LsraStat::REGSEL_FREE, refPosition->bbNum));
```
or is that too obfuscating?

**Review by BruceForstall — APPROVED** [review]

File: src/coreclr/jit/lsra.cpp (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/jit/lsra.cpp:3091]: I like the idea.

File: src/coreclr/jit/lsrastats.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+// Number of GT_COPY nodes inserted in this basic block while allocating regs.
+// Note that GT_COPY nodes are also inserted as part of basic block boundary
+// resolution, which are accounted against resolutionMovCount but not
+// against copyRegCount.
+LSRA_STAT_DEF(STAT_COPY_REG,                "CopyReg")
+
+// Number of resolution moves inserted in this basic block.
+LSRA_STAT_DEF(STAT_RESOLUTION_MOV,          "ResolutionMovs")
+
+// Number of critical edges from this block that are split.
```
[the remark this answers is not in the document]
Comment (kunalspathak) [reply src/coreclr/jit/lsrastats.h:26]: these are per block,  `LsraBlockInfo` to be precise.

**Comment by kunalspathak** [comment]
@BruceForstall  - I forgot to include `PerfScore` in LSRA stats (csv version). Since the perfscore are available only after code gen is complete, I had to add `dumpLsraStatsCsv()` method in `LinearScanInterface` so that I can call it from compiler.cpp. Please take another look and suggest if there is any better way to do it.

File: src/coreclr/jit/compiler.cpp [hunk 3]
```diff
@@ -5180,6 +5180,13 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
     // Generate code
     codeGen->genGenerateCode(methodCodePtr, methodCodeSize);
 
+#if TRACK_LSRA_STATS
+    if ((JitConfig.DisplayLsraStats() == 2))
```
Comment (BruceForstall) [inline src/coreclr/jit/compiler.cpp:5184]: nit: don't need double parens

File: src/coreclr/jit/lsra.cpp [hunk 4]
```diff
@@ -1284,14 +1291,18 @@ void LinearScan::doLinearScan()
...
-    if ((JitConfig.DisplayLsraStats() != 0)
+    if ((JitConfig.DisplayLsraStats() == 1)
 #ifdef DEBUG
         || VERBOSE
 #endif
         )
     {
         dumpLsraStats(jitstdout);
     }
+    else if ((JitConfig.DisplayLsraStats() == 2))
```
Comment (BruceForstall) [inline src/coreclr/jit/lsra.cpp:1302]: nit: don't need double parens

File: src/coreclr/jit/lsra.cpp [hunk 5]
```diff
@@ -3368,6 +3392,10 @@ regNumber LinearScan::allocateReg(Interval* currentInterval, RefPosition* refPos
         // We must have at least one with the lowest spill cost.
         assert(lowestCostSpillSet != RBM_NONE);
         found = selector.applySelection(SPILL_COST, lowestCostSpillSet);
+        if (found)
```
Comment (BruceForstall) [inline src/coreclr/jit/lsra.cpp:3395]: don't need the `if (found)`

File: src/coreclr/jit/lsra.cpp [hunk 6]
```diff
@@ -3398,6 +3426,10 @@ regNumber LinearScan::allocateReg(Interval* currentInterval, RefPosition* refPos
         // We must have at least one with the lowest spill cost.
         assert(farthestSet != RBM_NONE);
         found = selector.applySelection(FAR_NEXT_REF, farthestSet);
+        if (found)
```
Comment (BruceForstall) [inline src/coreclr/jit/lsra.cpp:3429]: don't need the `if (found)`

**Review by BruceForstall — APPROVED** [review]
LGTM modulo nits

File: src/coreclr/jit/lsra.cpp (hunk 4 above)
Comment (kunalspathak) [reply src/coreclr/jit/lsra.cpp:1302]: I actually forgot to remove this. We don't need `dumpLsraStatsCsv()` here.

**Pull request merged**